France-ioi/AlgoreaFrontend
# Permissions constraints

## Description

This is a redo of PR #692 

Fixes #593 

Permissions constraints are added through a custom validator in the permissions edition dialog box as suggested [here](https://github.com/France-ioi/AlgoreaFrontend/pull/692#issuecomment-925890649).
The values that are invalid due to permissions constraints are not grayed out according to the [constraints on receiver](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions). (for the `progress-select` and the `switch-field`)

### Testing

I wrote some unit tests for the permissions constraints custom validators [here](https://github.com/France-ioi/AlgoreaFrontend/pull/881/files#diff-04f985d45d516233c178c9906c8aa71307bcd517d4977add1d41551486b229f8)
For now every valid and invalid combinations of permissions are tested for the fields 'Can view' and 'Can grant view'.
Should I write the tests for the rest (the 4 other fields) ? or testing every cases is not useful ?

## Test cases

- [x] **Case 1**: You have the rights to edit the permissions of a user
  1. Given I am the user arbonenfant
  2. Start observing the groupe [UE Info - 2019-2020](https://dev.algorea.org/branch/permissions-constraints/en/#/groups/by-id/472375073987524072;path=/details)
  3. Go to the item [SSGeom](https://dev.algorea.org/branch/permissions-constraints/en/#/activities/by-id/1787514136807846970;path=4702,4102,1012286565380301759;attempId=0/details/progress/chapter) and in the Progress tab > Chapter View > Users, open the permission edit dialogue box for any users
  4. You can now edit permission (as an owner of the item, so almost all permissions)
  5. You can try changing permissions that depend on others, for example toggling 'Can attach official sessions' on and setting 'Can View' to 'Nothing' to invalidate the permissions

- [x] **Case 2**: You have the rights to edit the permissions of a user, but **not all permissions**
  1. Given I am the user arbonenfant
  2. Start observing the groupe [UE Info - 2019-2020](https://dev.algorea.org/branch/permissions-constraints/en/#/groups/by-id/472375073987524072;path=/details)
  3. Go to the item [SNT](https://dev.algorea.org/branch/permissions-constraints/en/#/activities/by-id/1352246428241737349;path=4702;attempId=0/details/progress/chapter) and in the Progress tab > Chapter View > Users, open the permission edit dialogue box for any users
  4. You can edit the permissions, but are very limited so you can see the permissions constraints at work.
